feat(ratings): support arbitrary weights for skill ratings [SCHOOL-24]

nbey opened the pull request:
Closes: SCHOOL-24

- feat: create migrations for new ratings
  - create `EvidenceWeight` column
  - migrate records to use `EvidenceWeight`
  - deprecate `Override` column
- feat: re-implement ability to create `Override` demonstrations
  - chore: remove deprecated field
  - fix: demonstration skill validation logic
  - fix: `SlateDemonstrationsTeacher` app Override creation

nbey committed:
feat: create migrations for new ratings

- create `EvidenceWeight` column
- migrate records to use `EvidenceWeight`
- deprecate `Override` column

nbey committed:
feat: create migrations for new ratings

- create `EvidenceWeight` column
- migrate records to use `EvidenceWeight`
- deprecate `Override` column

nbey committed:
feat: re-implement ability to create `Override` demonstrations

- chore: remove deprecated field
- fix: demonstration skill validation logic
- fix: `SlateDemonstrationsTeacher` app Override creation

nbey committed:
feat: re-implement ability to create `Override` demonstrations

- chore: remove deprecated field
- fix: demonstration skill validation logic
- fix: `SlateDemonstrationsTeacher` app Override creation

nbey committed:
wip: update student competency calculations

nbey committed:
wip: update student competency calculations

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -252,7 +254,8 @@ public function getDemonstrationData()
                             $demonstrationSkill['TargetLevel'] = intval($demonstrationSkill['TargetLevel']);
                             $demonstrationSkill['DemonstratedLevel'] = intval($demonstrationSkill['DemonstratedLevel']);
-                            $demonstrationSkill['Override'] = $demonstrationSkill['Override'] == '1';
+                            $demonstrationSkill['DemonstrationClass'] = $demonstrationSkill['DemonstrationClass'];
```
```suggestion
```

This line doesn't appear to be needed since there's no casting

themightychris committed:
refactor: refine override migrations

themightychris committed:
refactor: refine override migrations

themightychris committed:
fix: change EvidenceWeight default back to 1

themightychris committed:
fix: change EvidenceWeight default back to 1

nbey committed:
chore(docs): update student competency method docs

nbey committed:
chore(docs): update student competency method docs

chriscauley committed:
removing references to Vue

chriscauley committed:
app is running again

chriscauley committed:
converting the bulk of bootstrap components

chriscauley committed:
removing replacing dialog with vuetify component

chriscauley committed:
removing last of bootstrap references

chriscauley committed:
shifting slate-sidebar component into AdvancedPortfolioSidebar

chriscauley committed:
regrouping components (moving sidebar components into AdvancedPortfolioManager folder)

chriscauley committed:
rearranging css

chriscauley committed:
linting and cleaning up main.js

chriscauley committed:
fixing styles on skill demo card

themightychris committed:
chore: update package-lock.json

chriscauley committed:
fixing .block-loading

chriscauley committed:
sidebar was not connected to the x

chriscauley committed:
left a debug variable in

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -287,16 +289,29 @@ public function getDemonstrationOpportunities()
             }
         }
 
+        if ($hasWildCard) {
+            // return total requirements
+            $this->demonstrationOpportunities = $this->getDemonstrationsRequired();
+        }
```
Without this, could `$this->demonstrationOpportunities` ever be greater than `$this->getDemonstrationsRequired();` ? should this block handle capping it too if EvidenceWeight values > 1 are able to produce a higher count than required?

themightychris replied:
This doesn't look like it was previously working at all due to case mismatch in `$hasWildcard` vs `$hasWildCard`

themightychris replied:
fixed in d2b35a077f6be09cc64f996ee7ad97c326cc2093

themightychris commented on php-classes/Slate/CBL/Demonstrations/DemonstrationSkill.php:
```diff
@@ -35,9 +35,10 @@ class DemonstrationSkill extends \VersionedRecord
+            'type' => 'tinyint',
+            'unsigned' => true,
+            'default' => 1
         ]
```
- [ ] double-check the current codebase for any other instances of `Override` that might need to be removed/ported

themightychris committed:
docs: update CBL data model docs

themightychris committed:
docs: update CBL data model docs

themightychris committed:
docs: add example table

themightychris committed:
docs: add example table

themightychris committed:
test(e2e/api/student-competencies): update for EvidenceWeight

themightychris committed:
test(e2e/api/student-competencies): update for EvidenceWeight

themightychris committed:
fix: allow EvidenceWeight to be null

themightychris committed:
fix: allow EvidenceWeight to be null

themightychris committed:
fix: correct EvidenceWeight migration type

themightychris committed:
fix: correct EvidenceWeight migration type

themightychris committed:
fix: remove override-dependent validation of DemoSkill.DemonstratedLevel value

Demonstration skills will no longer know if they are part of an override and may have null demonstrated levels in any circumstance

themightychris committed:
fix: remove override-dependent validation of DemoSkill.DemonstratedLevel value

Demonstration skills will no longer know if they are part of an override and may have null demonstrated levels in any circumstance

themightychris committed:
fix: update saving of rating/weight data

themightychris committed:
fix: update saving of rating/weight data

themightychris committed:
chore: add diagnostic output to evidence weight migration

themightychris committed:
chore: add diagnostic output to evidence weight migration

themightychris committed:
test(e2e/api/student-competencies): add override save/delete tests

themightychris committed:
test(e2e/api/student-competencies): add override save/delete tests

themightychris committed:
fix: carry DemonstrationClass through data normalization

themightychris committed:
fix: carry DemonstrationClass through data normalization

themightychris committed:
fix: count missed demonstrations based only on DemonstratedLevel==0 and EvidenceWeight

themightychris committed:
fix: count missed demonstrations based only on DemonstratedLevel==0 and EvidenceWeight

themightychris committed:
test(e2e/api/student-competencies): add test deleting rating

themightychris committed:
test(e2e/api/student-competencies): add test deleting rating

themightychris committed:
test(e2e/api/student-competencies): use demonstration-skills API directly

themightychris committed:
test(e2e/api/student-competencies): use demonstration-skills API directly

themightychris committed:
docs: add description of overrides

themightychris committed:
docs: add description of overrides

themightychris committed:
fix: update HTML view of StudentCompetency details for new override model

themightychris committed:
fix: update HTML view of StudentCompetency details for new override model

themightychris committed:
chore: remove debug lines

themightychris committed:
chore: remove debug lines

themightychris committed:
style: add missing space

themightychris committed:
style: add missing space

themightychris committed:
test(e2e/api/student-competencies): rename test to ratings

themightychris committed:
test(e2e/api/student-competencies): rename test to ratings

themightychris committed:
fix(vscode): update studio ports for Chrome debuggers

themightychris committed:
fix(vscode): update studio ports for Chrome debuggers

themightychris committed:
feat: add weight and update override display in StudentCompetency detail page

themightychris committed:
feat: add weight and update override display in StudentCompetency detail page

themightychris committed:
test(e2e/api/ratings): re-organize by endpoint

themightychris committed:
test(e2e/api/ratings): re-organize by endpoint

themightychris committed:
test(e2e/api/ratings): add test setting DemonstratedLevel to null

themightychris committed:
test(e2e/api/ratings): add test setting DemonstratedLevel to null

themightychris committed:
test(e2e/api/ratings): add test ensuring ratings with no level/weight set are kept and not changed

themightychris committed:
test(e2e/api/ratings): add test ensuring ratings with no level/weight set are kept and not changed

themightychris committed:
test(e2e/api/ratings): remove frivelous IDs

themightychris committed:
test(e2e/api/ratings): remove frivelous IDs

themightychris committed:
fix(ratings): do not create new DemonstrationSkill records when no rating or weight provided

themightychris committed:
fix(ratings): do not create new DemonstrationSkill records when no rating or weight provided

themightychris committed:
test(e2e/api/ratings): add test for adding rating

themightychris committed:
test(e2e/api/ratings): add test for adding rating

themightychris committed:
fix(ratings): delete any DemoSkills without ratings after processing TaskSkills list

themightychris committed:
fix(ratings): delete any DemoSkills without ratings after processing TaskSkills list

themightychris committed:
fix: order StudentTask.Skills by ID

themightychris committed:
fix: order StudentTask.Skills by ID

themightychris committed:
test(e2e/api/ratings): add student-tasks ratings API tests

themightychris committed:
test(e2e/api/ratings): add student-tasks ratings API tests

themightychris committed:
fix(ratings): update bumping of Ms by updated overrides

themightychris committed:
fix(ratings): update bumping of Ms by updated overrides

themightychris committed:
docs: fix typo

themightychris committed:
docs: fix typo

themightychris committed:
fix: update recent-progress API for new override

themightychris committed:
fix: update recent-progress API for new override

themightychris committed:
fix: update student progress dashboard rendering of overrides

themightychris committed:
fix: update student progress dashboard rendering of overrides

themightychris committed:
feat: add demonstration opportunities to StudentCompetency detail page

themightychris committed:
feat: add demonstration opportunities to StudentCompetency detail page

themightychris committed:
fix: update calculating opportunities with wildcard/non-1 weights

themightychris committed:
fix: update calculating opportunities with wildcard/non-1 weights

themightychris committed:
fix: update teacher progress dashboard rendering of overrides

themightychris committed:
fix: update teacher progress dashboard rendering of overrides

themightychris committed:
fix: update skill history window rendering of overrides

themightychris committed:
fix: update skill history window rendering of overrides

themightychris committed:
fix(ratings): update demonstration sorting for new overrides

themightychris committed:
fix(ratings): update demonstration sorting for new overrides

themightychris committed:
fix: update legacy MRMF growth calculator for new overrides schema

themightychris committed:
fix: update legacy MRMF growth calculator for new overrides schema

themightychris committed:
fix: update demonstrations exporter for new overrides schema

themightychris committed:
fix: update demonstrations exporter for new overrides schema

themightychris committed:
fix: update Sencha fields/data for new overrides schema

themightychris committed:
fix: update Sencha fields/data for new overrides schema

themightychris committed:
fix(portfolio-manager): update rendering for new overrides schema

themightychris committed:
fix(portfolio-manager): update rendering for new overrides schema

themightychris committed:
test(e2e/api/ratings): remove hardcoded demoskill ID

themightychris committed:
test(e2e/api/ratings): remove hardcoded demoskill ID

ryon committed:
feat: overhaul styling and layout

chriscauley committed:
progress was not properly hooked up to LevelPanel

themightychris committed:
Merge pull request #754 from SlateFoundation/issue/742-vue3-and-vuetify

Issue/742 vue3 and vuetify

themightychris committed:
chore(deps): bump slate to v2.21.0

themightychris committed:
feat(portfolio-manager): set up custom frame+loader

themightychris committed:
fix(portfolio-manager): counteract Slate frame CSS

themightychris committed:
feat(portfolio-manager): add to tools menu

themightychris committed:
chore(deps): bump slate to v2.21.1

themightychris committed:
fix(portfolio-manager): remove external font references

themightychris committed:
chore(deps): bump slate to v2.21.2

themightychris committed:
feat(k8s): enable wildcard CORS for deployments

themightychris committed:
fix(tasks-teacher): restore Created sort after archived sort

chriscauley committed:
Pop-up to edit baseline

chriscauley committed:
info popup

themightychris committed:
Merge pull request #774 from SlateFoundation/fixes/teacher-tasks-sort

fix(tasks-teacher): restore Created sort after archived sort

themightychris committed:
Merge pull request #775 from SlateFoundation/issue-737/editing-baseline

Pop-up to edit baseline

themightychris committed:
Merge pull request #776 from SlateFoundation/issue-735/info-popup

info popup

BillClinton committed:
fix: wait for demonstrations to load before click

BillClinton committed:
fix: wait for demonstrations to load before click

BillClinton committed:
fix: wait for XHR before checking slider component

BillClinton committed:
fix: wait for XHR before checking slider component

BillClinton committed:
Merge branch 'fix/override-form-race-condition' of github.com:SlateFoundation/slate-cbl into fix/override-form-race-condition

BillClinton committed:
fix: correct race conditions in final test

themightychris committed:
fix(e2e): remove errant .only

BillClinton committed:
fix: avoid extGet in `within` context

themightychris committed:
fix(e2e): eliminate lingering unscoped cy.extGet

themightychris committed:
Merge pull request #777 from SlateFoundation/fix/override-form-race-condition

fix(e2e): prevent race conditions during overrides tests

themightychris committed:
fix(tools): only show enrollments dashboard to admins

themightychris committed:
refactor: move CBL defaults upstream

themightychris committed:
chore(deps): bump slate to v2.21.3

themightychris committed:
fix(e2e/progress/overrides): correct case for witheld graduation

themightychris committed:
fix(graduation): skip minimum average requirement when no average is available

BillClinton committed:
refactor: avoid extGet in withing block

chriscauley committed:
skill-demo-card classes got disconnected

chriscauley committed:
removing todos and adding DemonstrationSkillID

BillClinton committed:
fix: don't use within when scope is used only once

BillClinton committed:
fix: retain `extGet` calls but move outside within

BillClinton committed:
fix: correct race condition in column sorting test

themightychris committed:
Merge pull request #781 from SlateFoundation/fixes/refactor-task-library-tests

refactor: avoid extGet in within block

themightychris committed:
Merge pull request #782 from SlateFoundation/ccc-fixes

Minor Fixes to Advanced Profile Sidebar

BillClinton committed:
fix: show previous portfolios for enrollment

BillClinton committed:
refactor: remove comment after verifying

BillClinton committed:
refactor: update variable name

enrolledLevels is a more appropriate variable name than visibleLevels now that
all levels are visible regardless of enrolled status

themightychris committed:
Merge pull request #786 from SlateFoundation/fixes/allow-creation-of-previous-portfolios

fix: show previous portfolios for enrollment

themightychris committed:
refactor(ci): use upstream release workflows

themightychris committed:
chore(dev): update studio script

themightychris committed:
refactor(dev): use studio base image container

themightychris committed:
chore(deps): bump slate to v2.21.4

themightychris committed:
refactor(ci): use upstream container and preview workflows

themightychris committed:
chore(ci): delete deprecated container build assets

themightychris committed:
refactor(docs): move mkdocs config under docs/

themightychris committed:
fix(ci): use bot token to prepare PR

BillClinton committed:
Merge branch 'feat/ratings-backend' of github.com:SlateFoundation/slate-cbl into feat/ratings-backend
